File: app/src/main/java/com/stevenfrew/beatprompter/ChooseFolderDialog.java
package com.stevenfrew.beatprompter;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.view.Window;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;

import java.util.List;
import java.util.Locale;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

abstract class ChooseFolderDialog implements DialogInterface.OnCancelListener,DialogInterface.OnDismissListener {
    private static final String PARENT_DIR = "..";

    Handler mChooseFolderDialogHandler = new Handler() {
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case BeatPrompterApplication.FOLDER_CONTENTS_FETCHED:
                    populateBrowser((List<CloudBrowserItem>)msg.obj);
                    break;
                case BeatPrompterApplication.FOLDER_CONTENTS_FETCHING:
                    updateProgress(msg.arg1,msg.arg2);
                    break;
            }
        }
    };

    private Dialog mDialog;
    private CloudBrowserItem mCurrentFolder;
    private CloudBrowserItem mParentFolder;
    private Activity mActivity;
    private FolderFetcherTask mFolderFetcher;
    private PublishSubject<CloudFolderInfo> mFolderSelectionSource=PublishSubject.create();

    ChooseFolderDialog(final Activity activity,int iconResourceID) {
        mActivity=activity;
        mDialog = new Dialog(activity,R.style.CustomDialog);
        mDialog.requestWindowFeature(Window.FEATURE_LEFT_ICON);
        mDialog.setContentView(R.layout.choose_folder_dialog_loading);
        mDialog.setCanceledOnTouchOutside(false);
        mDialog.setOnCancelListener(this);
        mDialog.setOnDismissListener(this);
        TextView tv = mDialog.findViewById(android.R.id.title);
        tv.setEllipsize(TextUtils.TruncateAt.END);
        tv.setLines(1);
        tv.setHorizontallyScrolling(true);
        mDialog.setFeatureDrawableResource(Window.FEATURE_LEFT_ICON, iconResourceID);
    }

    void showDialog()
    {
        mCurrentFolder=getRootPath();
        if(mCurrentFolder==null)
            return;
        refresh(mCurrentFolder);
        mDialog.setTitle(getDisplayPath(mCurrentFolder));
        mDialog.show();
    }

    private void setNewPath(final CloudBrowserItem newFolder) {
        if(newFolder!=null)
            mFolderSelectionSource.onNext(new CloudFolderInfo(newFolder.mInternalPath, getDisplayPath(newFolder)));
        mDialog.dismiss();
    }

    private void refresh(CloudBrowserItem folder)
    {
        if(folder==null)
            return;
        this.mCurrentFolder = folder;
        this.mParentFolder=folder.mParentFolder;
        cancelFolderFetcher();
        mFolderFetcher=getFolderFetcher();
        mFolderFetcher.execute(mCurrentFolder);
    }

    private void populateBrowser(List<CloudBrowserItem> dirs) {
        if (dirs == null)
            mDialog.dismiss();
        else {
            dirs.sort(CloudBrowserItem::compareTo);

            if (mCurrentFolder.mParentFolder != null)
                dirs.add(0, new CloudBrowserItem(mCurrentFolder.mParentFolder, PARENT_DIR, mCurrentFolder.mParentFolder.mInternalPath, true));

            // refresh the user interface
            mDialog.setContentView(R.layout.choose_folder_dialog);
            final ListView list = mDialog.findViewById(R.id.chooseFolderListView);
            Button okButton = mDialog.findViewById(R.id.chooseFolderOkButton);
            mDialog.setTitle(getDisplayPath(mCurrentFolder));
            list.setAdapter(new CloudBrowserItemListAdapter(dirs));

            list.setOnItemClickListener((parent, view, which, id) -> {
                CloudBrowserItem folderChosen;
                if((which==0)&&(mParentFolder!=null))
                    folderChosen=mParentFolder;
                else
                    folderChosen = (CloudBrowserItem) list.getItemAtPosition(which);

                mDialog.setContentView(R.layout.choose_folder_dialog_loading);
                mDialog.setTitle(getDisplayPath(folderChosen));
                refresh(folderChosen);
            });
            okButton.setOnClickListener(v -> setNewPath(mCurrentFolder));
        }
    }

    abstract CloudBrowserItem getRootPath();
    abstract String getDirectorySeparator();
    abstract ChooseFolderDialog.FolderFetcherTask getFolderFetcher();

    private String getDisplayPath(CloudBrowserItem folder)
    {
        if(folder.mParentFolder!=null) {
            String parentPath=getDisplayPath(folder.mParentFolder);
            if(!parentPath.endsWith(getDirectorySeparator()))
                parentPath+=getDirectorySeparator();
            return parentPath + folder.mDisplayName;
        }
        return folder.mDisplayName;
    }

    private void updateProgress(int found,int max)
    {
        TextView progressText = (TextView) mDialog.findViewById(R.id.loading_count);
        if(progressText!=null)
            if(max==0) {
                String itemsFound = mActivity.getString(R.string.itemsFound);
                progressText.setText(String.format(Locale.getDefault(), itemsFound, found));
            }
            else
            {
                String itemsFound = mActivity.getString(R.string.itemsFoundWithMax);
                progressText.setText(String.format(Locale.getDefault(), itemsFound, found, max));
            }
    }

    public void onDismiss(DialogInterface dialog)
    {
        cancelFolderFetcher();
        mFolderSelectionSource.onComplete();
    }

    public void onCancel(DialogInterface dialog)
    {
        cancelFolderFetcher();
        mFolderSelectionSource.onComplete();
    }

    private void cancelFolderFetcher()
    {
        if(mFolderFetcher!=null) {
            mFolderFetcher.cancel(true);
            mFolderFetcher=null;
        }
    }

    abstract class FolderFetcherTask extends AsyncTask<CloudBrowserItem, Void, List<CloudBrowserItem>>
    {
        protected abstract List<CloudBrowserItem> doInBackground(CloudBrowserItem... args);
        protected void onPostExecute(List<CloudBrowserItem> folders)
        {
            mChooseFolderDialogHandler.obtainMessage(BeatPrompterApplication.FOLDER_CONTENTS_FETCHED,folders).sendToTarget();
        }
    }

    Observable<CloudFolderInfo> getFolderSelectionSource()
    {
        return mFolderSelectionSource;
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/storage/ChooseFolderDialog.kt
package com.stevenfrew.beatprompter.storage

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Handler
import android.os.Message
import android.text.TextUtils
import android.view.Window
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import com.stevenfrew.beatprompter.R
import com.stevenfrew.beatprompter.cache.Cache
import com.stevenfrew.beatprompter.events.Events
import com.stevenfrew.beatprompter.util.CoroutineTask
import com.stevenfrew.beatprompter.util.Utils
import com.stevenfrew.beatprompter.util.execute
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import java.util.Locale
import kotlin.coroutines.CoroutineContext

/**
 * Dialog allowing the user to choose the folder to use as the source for data.
 */
internal class ChooseFolderDialog(
	private val activity: Activity,
	private val storage: Storage,
	listener: FolderSelectionListener,
	private var currentFolder: FolderInfo
) : DialogInterface.OnCancelListener, DialogInterface.OnDismissListener, FolderSearchListener {
	private var folderSearchError: Throwable? = null
	private val dialog: Dialog = Dialog(activity, R.style.FolderBrowserDialog)
	private val folderSelectionEventSubscription: CompositeDisposable
	private var parentFolder: FolderInfo? = null
	private val handler: FolderContentsFetchHandler
	private var folderFetcher: FolderFetcherTask? = null
	private val folderSelectionSource = PublishSubject.create<FolderInfo?>()
	private val displayItems = ArrayList<ItemInfo>()
	private var shouldCancel: Boolean = false

	internal class FolderContentsFetchHandler(private var chooseFolderDialog: ChooseFolderDialog) :
		Handler() {
		override fun handleMessage(msg: Message) {
			@Suppress("UNCHECKED_CAST")
			when (msg.what) {
				Events.FOLDER_CONTENTS_FETCHED -> chooseFolderDialog.populateBrowser(msg.obj as MutableList<ItemInfo>)
				Events.FOLDER_CONTENTS_FETCHING -> chooseFolderDialog.updateProgress(msg.arg1, msg.arg2)
			}
		}
	}

	init {
		val thisDialog = this
		dialog.apply {
			requestWindowFeature(Window.FEATURE_LEFT_ICON)
			setContentView(R.layout.choose_folder_dialog_loading)
			setCanceledOnTouchOutside(false)
			setFeatureDrawableResource(Window.FEATURE_LEFT_ICON, storage.cloudIconResourceId)
			dialog.setOnCancelListener(thisDialog)
			dialog.setOnDismissListener(thisDialog)
		}

		handler = FolderContentsFetchHandler(this)
		folderSelectionEventSubscription = CompositeDisposable().apply {
			add(
				folderSelectionSource.subscribe(
					{ listener.onFolderSelected(it) },
					{ listener.onFolderSelectedError(it, activity) },
					{ listener.onFolderSelectionComplete() })
			)
		}

		dialog.findViewById<TextView>(android.R.id.title).apply {
			ellipsize = TextUtils.TruncateAt.END
			setLines(1)
			setHorizontallyScrolling(true)
		}
	}

	fun showDialog() {
		refresh(currentFolder)
		dialog.apply {
			setTitle(getDisplayPath(currentFolder))
			show()
		}
	}

	private fun setNewPath(newFolder: FolderInfo) {
		folderSelectionSource.onNext(newFolder)
		dialog.dismiss()
	}

	private fun refresh(folder: FolderInfo?) {
		if (folder == null)
			return
		currentFolder = folder
		parentFolder = folder.parentFolder
		cancelFolderFetcher()
		displayItems.clear()
		folderFetcher = FolderFetcherTask(storage, this)
		folderFetcher!!.execute(currentFolder)
	}

	private fun populateBrowser(contents: MutableList<ItemInfo>?) {
		if (contents == null)
			dialog.dismiss()
		else {
			contents.removeAll(Cache.defaultDownloads.map { it.fileInfo })
			contents.sort()

			currentFolder.parentFolder?.also {
				contents.add(0, FolderInfo(it.parentFolder, it.id, PARENT_DIR, it.displayPath))
			}

			// refresh the user interface
			dialog.apply {
				setContentView(R.layout.choose_folder_dialog)
				setTitle(getDisplayPath(currentFolder))
				val list = findViewById<ListView>(R.id.chooseFolderListView)
				list.adapter = BrowserItemListAdapter(contents, context)

				list.setOnItemClickListener { _, _, which, _ ->
					val folderChosen: FolderInfo = if (which == 0 && parentFolder != null)
						parentFolder as FolderInfo
					else
						list.getItemAtPosition(which) as FolderInfo

					setContentView(R.layout.choose_folder_dialog_loading)
					setTitle(getDisplayPath(folderChosen))
					refresh(folderChosen)
				}
				val okButton = findViewById<Button>(R.id.chooseFolderOkButton)
				if (folderSearchError == null)
					okButton.setOnClickListener { setNewPath(currentFolder) }
				else
					okButton.isEnabled = false
			}
		}
	}

	private fun getDisplayPath(folder: FolderInfo): String {
		if (folder.parentFolder != null) {
			var parentPath = getDisplayPath(folder.parentFolder)
			if (!parentPath.endsWith(storage.directorySeparator))
				parentPath += storage.directorySeparator
			return parentPath + folder.name
		}
		return folder.name
	}

	private fun updateProgress(found: Int, max: Int) {
		val progressText = dialog.findViewById<TextView>(R.id.loading_count)
		if (progressText != null)
			if (max == 0) {
				val itemsFound = activity.getString(R.string.itemsFound)
				progressText.text = String.format(Locale.getDefault(), itemsFound, found)
			} else {
				val itemsFound = activity.getString(R.string.itemsFoundWithMax)
				progressText.text = String.format(Locale.getDefault(), itemsFound, found, max)
			}
	}

	override fun onDismiss(dialog: DialogInterface) {
		cancelFolderFetcher()
		folderSelectionSource.onComplete()
		folderSelectionEventSubscription.dispose()
	}

	override fun onCancel(dialog: DialogInterface) {
		onDismiss(dialog)
	}

	private fun cancelFolderFetcher() {
		if (folderFetcher != null)
			try {
				folderFetcher!!.cancel("Cancelling folder fetcher ...")
			} catch (_: IllegalStateException) {
				// Fetcher wasn't running, so nothing to cancel.
			}
		folderFetcher = null
	}

	class FolderFetcherTask internal constructor(
		private val storage: Storage,
		private val folderSearchListener: FolderSearchListener
	) : CoroutineTask<FolderInfo, Unit, Unit> {
		override val coroutineContext: CoroutineContext
			get() = Dispatchers.IO

		override fun onPreExecute() {
			// Do nothing.
		}

		override fun onError(t: Throwable) {
			// Listener will receive error callbacks.
		}

		override fun onProgressUpdate(progress: Unit) {
			// No progress updating required.
		}

		override fun onPostExecute(result: Unit) {
			// Do nothing. Listener will receive a completion callback.
		}

		override fun doInBackground(params: FolderInfo, progressUpdater: suspend (Unit) -> Unit) =
			storage.readFolderContents(params, folderSearchListener, recurseSubFolders = false)
	}

	override fun onCloudItemFound(item: ItemInfo) {
		displayItems.add(item)
	}

	override fun onFolderSearchError(t: Throwable, context: Context) {
		activity.runOnUiThread {
			Utils.showExceptionDialog(t, context)
		}
		folderSearchError = t
		onFolderSearchComplete()
	}

	override fun onFolderSearchComplete() {
		displayItems.sort()
		handler.obtainMessage(Events.FOLDER_CONTENTS_FETCHED, displayItems).sendToTarget()
	}

	override suspend fun onProgressMessageReceived(message: String) {
		// Do nothing.
	}

	override fun shouldCancel(): Boolean = shouldCancel

	companion object {
		private const val PARENT_DIR = ".."
	}
}